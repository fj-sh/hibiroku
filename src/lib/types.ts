export type Task = {
  id: string
  name: string
  status: TimerStatus
  totalSeconds: number
}

export type TimerStatus = 'inProgress' | 'Stopped'
