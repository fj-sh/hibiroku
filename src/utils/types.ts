type Task = {
  id: string
  name: string
  timer: Timer
}

type Timer = {
  total: string
  timeStatus: TimerStatus
}

type TimerStatus = 'inProgress' | 'Stopped'
