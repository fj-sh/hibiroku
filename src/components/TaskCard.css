.TaskCard_container {
  display: flex;
  flex-direction: column;
  width: 95%;
  height: auto;
  margin: 0 auto 1.5rem auto;
  background: #ffffff;
  border-radius: 10px;
  box-shadow: rgba(0, 0, 0, 0.08) 0px 4px 12px;
}

.TaskCard_task_name_seconds_wrapper {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin: 0.2rem 2rem 0.2rem 2rem;
}

.TaskCard_task_name {
  font-size: 18px;
  text-align: center;
  align-content: center;
  font-weight: 900;
}
.TaskCard_task_seconds {
  font-size: 18px;
  text-align: center;
  align-content: center;
  font-weight: 900;
}

.TaskCard_task_status {

  font-size: 12px;
  margin: 0.3rem auto 0.3rem auto;
  align-items: center;
}


@keyframes fade {
  from { opacity: 1.0; }
  50% { opacity: 0.5; }
  to { opacity: 1.0; }
}

@-webkit-keyframes fade {
  from { opacity: 1.0; }
  50% { opacity: 0.5; }
  to { opacity: 1.0; }
}

.TaskCard_task_status_inProgress {
  background: #80ff96;
  color: #000000;
  border-radius: 5px;
  width: 100px;
  font-size: 12px;
  font-weight: 700;
  padding: 0.2rem;
  margin: 0.3rem 0 0.3rem 0;
  text-align: center;
  animation:fade 3000ms infinite;
  -webkit-animation:fade 3000ms infinite;
}

.TaskCard_task_status_pending {
  background: firebrick;
  color: #eaeaea;
  border-radius: 5px;
  width: 100px;
  font-size: 12px;
  font-weight: 700;
  padding: 0.2rem;
  margin: 0.3rem 0 0.3rem 0;
  text-align: center;
}

.TaskCard_button_wrapper {
  display: flex;
  flex-direction: row;
  margin: 0 1.5rem 1rem 1.5rem;
  justify-content: space-between;
}

.TaskCard_button {
  display: inline-block;
  padding: 0.3em 1.2em;
  margin: 0 0.1em 0.1em 0;
  border: 0.16em solid rgba(255,255,255,0);
  border-radius: 2em;
  box-sizing: border-box;
  text-decoration: none;
  font-weight: 300;
  color: #FFFFFF;
  text-shadow: 0 0.04em 0.04em rgba(0,0,0,0.35);
  text-align: center;
  transition: all 0.2s;
}

.TaskCard_button:hover {
  border-color: rgba(255,255,255,1);
}

@media all and (max-width:30em){
  .TaskCard_button {
    display:block;
    margin:0.2em auto;
  }
}

.TaskCard_button_start {
  background-color:#4e9af1
}


.TaskCard_button_stop {
  background-color:#f1bb4e
}

.TaskCard_button_reset {
  background-color:#9a4ef1
}

.TaskCard_button_delete {
  background-color:#f14e4e
}

