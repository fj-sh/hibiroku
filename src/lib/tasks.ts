import { Task, TimerStatus } from './types'

export const getTasksFromStorage = async (): Promise<Task[]> => {
  const res = await chrome.storage.local.get(['tasks'])
  return res.tasks as Task[]
}

export const getTaskFromStorage = async (taskId: string) => {
  const tasksFromStorage = await getTasksFromStorage()
  if (tasksFromStorage.length === 0) return undefined
  return tasksFromStorage.filter((task) => task.id === taskId)[0]
}

export const persistTasksInChromeStorage = (tasks: Task[]) => {
  chrome.storage.local.set(
    {
      tasks: tasks,
    },
    () => {
      /* [Callback] Task saved in Chrome local storage. */
    }
  )
}

export const toggleTaskStatus = async (taskId: string): Promise<TimerStatus> => {
  const task = await getTaskFromStorage(taskId)
  const updatedStatus: TimerStatus = task.status === 'inProgress' ? 'Stopped' : 'inProgress'
  const updatedTask = {
    ...task,
    status: updatedStatus,
  }
  updateTaskInChromeStorage(updatedTask)
  return updatedStatus
}

export const getCurrentStatus = async (taskId: string): Promise<TimerStatus> => {
  const task = await getTaskFromStorage(taskId)
  return task.status
}

export const addTaskInChromeStorage = async (task: Task) => {
  const tasks = await getTasksFromStorage()
  const updatedTasks = [...tasks, task]
  persistTasksInChromeStorage(updatedTasks)
}

export const updateTaskInChromeStorage = (task: Task) => {
  chrome.storage.local.get(['tasks'], (res) => {
    const storedTasks: Task[] = res.tasks ?? []
    if (storedTasks.length === 0) return
    const updatedTasks = storedTasks.map((storedTask) => {
      if (storedTask.id === task.id) {
        return task
      }
      return storedTask
    })
    persistTasksInChromeStorage(updatedTasks)
  })
}

export const deleteTaskInChromeStorage = async (taskId: string) => {
  const storedTasks = await getTasksFromStorage()
  const updatedTasks = storedTasks.filter((storedTask) => storedTask.id !== taskId)
  persistTasksInChromeStorage(updatedTasks)
}

export const addOneSecondToTasksInProgress = () => {
  chrome.storage.local.get(['tasks'], (res) => {
    const storedTasks: Task[] = res.tasks ?? []
    if (storedTasks.length === 0) return
    const updatedTasks = storedTasks.map((storedTask) => {
      if (!('status' in storedTask)) return
      if (storedTask.status === 'inProgress') {
        return { ...storedTask, totalSeconds: storedTask.totalSeconds + 1 }
      } else {
        return storedTask
      }
    })
    persistTasksInChromeStorage(updatedTasks)
  })
}

export const resetTimer = async (taskId: string) => {
  const task = await getTaskFromStorage(taskId)
  task.totalSeconds = 0
  updateTaskInChromeStorage(task)
  return task
}

export const resetAllTasks = async () => {
  const storedTasks = await getTasksFromStorage()
  const updatedTasks = storedTasks.map((storedTask) => {
    return {
      ...storedTask,
      totalSeconds: 0,
    }
  })
  persistTasksInChromeStorage(updatedTasks)
}
