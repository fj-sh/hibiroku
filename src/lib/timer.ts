export const secondsToHms = (totalSeconds: number) => {
  const hour = Math.floor(totalSeconds / 3600)
  const minute = Math.floor((totalSeconds % 3600) / 60)
  const seconds = Math.floor((totalSeconds % 3600) % 60)

  const hourDisplay = hour > 0 ? String(hour).padStart(2, '0') : '00'
  const minuteDisplay = minute > 0 ? String(minute).padStart(2, '0') : '00'
  const secondsDisplay = seconds > 0 ? String(seconds).padStart(2, '0') : '00'
  return `${hourDisplay}:${minuteDisplay}:${secondsDisplay}`
}
