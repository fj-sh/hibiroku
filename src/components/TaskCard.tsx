import React, { useEffect, useState } from 'react'
import { Task, TimerStatus } from '../lib/types'
import { secondsToHms } from '../lib/timer'
import { getCurrentStatus, getTaskFromStorage, resetTimer, toggleTaskStatus } from '../lib/tasks'
import './TaskCard.css'

type Props = {
  task: Task
  onDelete: (taskId: string) => void
}

const TaskCard = ({ task, onDelete }: Props) => {
  const [seconds, setSeconds] = useState<number>(0)
  const [status, setStatus] = useState<TimerStatus>('Stopped')

  let timer = undefined
  const updateSeconds = () => {
    timer = setInterval(async () => {
      const storedTask = await getTaskFromStorage(task.id)
      setSeconds(storedTask.totalSeconds)
    }, 1000)
  }

  const setCurrentStatus = (taskId: string) => {
    getCurrentStatus(taskId).then((status) => setStatus(status))
  }

  useEffect(() => {
    setCurrentStatus(task.id)
    updateSeconds()
    return () => clearInterval(timer)
  }, [])

  const handleToggleStatus = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    task: Task
  ) => {
    toggleTaskStatus(task.id).then((updatedStatus) => {
      setStatus(updatedStatus)
    })
  }

  const handleResetTimer = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>, task: Task) => {
    resetTimer(task.id).then((task) => {
      setSeconds(task.totalSeconds)
    })
  }

  const handleDeleteTask = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    taskId: string
  ) => {
    clearInterval(timer)
    onDelete(taskId)
  }

  return (
    <div className="TaskCard_container">
      <label className="TaskCard_task_status">
        {status === 'inProgress' ? (
          <div className="TaskCard_task_status_inProgress">In Progress</div>
        ) : (
          <div className="TaskCard_task_status_pending">Pending</div>
        )}
      </label>
      <div className="TaskCard_task_name_seconds_wrapper">
        <h2 className="TaskCard_task_name">{task.name}</h2>
        <time className="TaskCard_task_seconds">{secondsToHms(seconds)}</time>
      </div>

      <div className="TaskCard_button_wrapper">
        <button
          className={
            status === 'inProgress'
              ? 'TaskCard_button TaskCard_button_stop'
              : 'TaskCard_button TaskCard_button_start'
          }
          onClick={(event) => handleToggleStatus(event, task)}
        >
          {status === 'inProgress' ? <>Stop</> : <>Start</>}
        </button>

        <button
          className="TaskCard_button TaskCard_button_reset"
          onClick={(event) => handleResetTimer(event, task)}
        >
          Reset
        </button>
        <button
          className="TaskCard_button TaskCard_button_delete"
          onClick={(event) => handleDeleteTask(event, task.id)}
        >
          Delete
        </button>
      </div>
    </div>
  )
}

export default TaskCard
