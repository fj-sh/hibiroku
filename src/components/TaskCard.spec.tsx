import React from 'react'
import renderer from 'react-test-renderer'
import TaskCard from './TaskCard'
import { Task } from '../lib/types'

describe('TaskCard', () => {
  it('Create snapshot', () => {
    const task: Task = {
      id: 'aaa',
      name: 'task',
      status: 'Stopped',
      totalSeconds: 1,
    }
    const onDelete = () => {}

    const component = renderer.create(<TaskCard task={task} onDelete={onDelete} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
