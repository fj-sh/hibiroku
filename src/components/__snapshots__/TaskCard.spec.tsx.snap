// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCard Create snapshot 1`] = `
Array [
  "task",
  <button
    onClick={[Function]}
  >
    Start
  </button>,
  "00:00:00",
  <button
    onClick={[Function]}
  >
    Reset
  </button>,
  <button
    onClick={[Function]}
  >
    Delete
  </button>,
  <br />,
]
`;
